pkg/operator/controller/controller_dns_daemonset.go

```diff
@@ -0,0 +1,186 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/cluster-dns-operator/pkg/manifests"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+)
+
+// ensureDNSDaemonSet ensures the dns daemonset exists for a given cluster dns.
+func (r *reconciler) ensureDNSDaemonSet(dns *operatorv1.DNS, clusterIP, clusterDomain string) (*appsv1.DaemonSet, error) {
+	desired, err := r.desiredDNSDaemonSet(dns, clusterIP, clusterDomain)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build dns daemonset: %v", err)
+	}
+	current, err := r.currentDNSDaemonSet(dns)
+	if err != nil {
+		return nil, err
+	}
+	switch {
+	case desired != nil && current == nil:
+		if err := r.createDNSDaemonSet(desired); err != nil {
+			return nil, err
+		}
+	case desired != nil && current != nil:
+		if err := r.updateDNSDaemonSet(current, desired); err != nil {
+			return nil, err
+		}
+	}
+	return r.currentDNSDaemonSet(dns)
+}
+
+// ensureDNSDaemonSetDeleted ensures deletion of daemonset and related resources
+// associated with the cluster dns.
+func (r *reconciler) ensureDNSDaemonSetDeleted(dns *operatorv1.DNS) error {
+	daemonset := &appsv1.DaemonSet{}
+	name := DNSDaemonSetName(dns)
+	daemonset.Name = name.Name
+	daemonset.Namespace = name.Namespace
+	if err := r.client.Delete(context.TODO(), daemonset); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+	}
+	logrus.Infof("deleted dns daemonset: %s", dns.Name)
```

Comment:
I kind of feel like this log message should be in the `else` for the preceding `if`:
```go
	if err := r.client.Delete(context.TODO(), daemonset); err != nil {
		if !errors.IsNotFound(err) {
			return err
		}
	} else {
		logrus.Infof("deleted dns daemonset: %s", dns.Name)
	}
	return nil
```
What do you think?

Reply:
done